src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -161,4 +198,102 @@ public override BoundNode VisitAwaitableValuePlaceholder(BoundAwaitableValuePlac
+                visitedRight,
+                _proxies,
+                createHoistedLocal,
+                createHoistedAccess,
+                this,
+                isRuntimeAsync: true);
+        }
+
+        var visitedLeftOrProxy = VisitExpression(leftLocal);
+        visitedRight = VisitExpression(node.Right);
+
+        if (visitedLeftOrProxy is not BoundLocal visitLeftLocal)
+        {
+            // Proxy replacement occurred. We need to reassign the proxy into our local as a sequence.
+            // ref leftLocal = ref proxy;
+            // leftLocal = visitedRight;
+            var assignment = _factory.AssignmentExpression(leftLocal, visitedLeftOrProxy, isRef: true);
+            return _factory.Sequence([assignment], node.Update(leftLocal, visitedRight, node.IsRef, node.Type));
+        }
+
+        return node.Update(visitedLeftOrProxy, visitedRight, node.IsRef, node.Type);
+
+        static LocalSymbol createHoistedLocal(TypeSymbol type, RuntimeAsyncRewriter @this, LocalSymbol local)
+        {
+            var hoistedLocal = @this._factory.SynthesizedLocal(type, syntax: local.GetDeclaratorSyntax(), kind: SynthesizedLocalKind.AwaitByRefSpill);
+            @this._hoistedLocals.Add(hoistedLocal);
+            return hoistedLocal;
+        }
+
+        static BoundLocal createHoistedAccess(LocalSymbol local, RuntimeAsyncRewriter @this)
+            => @this._factory.Local(local);
+    }
+
+    private bool TryReplaceWithProxy(Symbol localOrParameter, SyntaxNode syntax, [NotNullWhen(true)] out BoundNode? replacement)
+    {
+        if (_proxies.TryGetValue(localOrParameter, out CapturedSymbolReplacement? proxy))
+        {
+            replacement = proxy.Replacement(syntax, makeFrame: null, this);
+            return true;
+        }
+
+        replacement = null;
+        return false;
+    }
+
+    public sealed override BoundNode VisitLocal(BoundLocal node)
+    {
+        if (TryReplaceWithProxy(node.LocalSymbol, node.Syntax, out BoundNode? replacement))
+        {
+            return replacement;
+        }
+
+        Debug.Assert(!_variablesToHoist.Contains(node.LocalSymbol));
+        return base.VisitLocal(node)!;
+    }
+
+    public override BoundNode? VisitParameter(BoundParameter node)
+    {
+        if (TryReplaceWithProxy(node.ParameterSymbol, node.Syntax, out BoundNode? replacement))
+        {
+            // Currently, the only parameter we expect to be replaced is `this`, which is handled through VisitThisReference.
+            // Any other ref to a parameter should have either already been hoisted to a local during local rewriting, or should
+            // be an illegal ref to a parameter across an await.
+            throw ExceptionUtilities.Unreachable();
+        }
+
+        Debug.Assert(!_variablesToHoist.Contains(node.ParameterSymbol));
+        return base.VisitParameter(node);
+    }
+
+    public override BoundNode? VisitThisReference(BoundThisReference node)
```

Comment:
Looking at `RuntimeAsyncRewriter`'s usage of `TryReplaceWithProxy`, I see three cases: locals, parameters and `this`. But in `MethodToClassRewriter`, I only see locals and parameters. I'm not sure why. Do you understand why?